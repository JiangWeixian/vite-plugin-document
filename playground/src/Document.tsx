import React from 'react'

const Document = () => {
  return (
    <html lang="en">
      <head>
        <meta charSet="UTF-8" />
        <link rel="icon" type="image/svg+xml" href="/src/favicon.svg" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>Vite App</title>
      </head>
      <body>
        <div id="root"></div>
        <script type="module" src="/src/main.tsx"></script>
      </body>
    </html>
  )
}

export default Document
