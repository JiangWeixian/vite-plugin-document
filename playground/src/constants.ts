export const title = 'Vite App'
