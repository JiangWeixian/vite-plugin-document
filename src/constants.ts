export const VIRTUAL_DOCUMENT_ID = 'virtual-document.tsx'
export const VIRTUAL_SOLID_HTML_ID = 'virtual-solid-html.tsx'
export const VIRTUAL_REACT_HTML_ID = 'virtual-react-html.tsx'
export const NAME = 'vite-plugin-document'
