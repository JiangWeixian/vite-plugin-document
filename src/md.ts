import { table } from 'functional-md'

export const Table = () => {
  return table({
    columns: [
      {
        dataIndex: 'id',
        title: 'ID',
      },
      {
        dataIndex: '',
        title: 'NAME',
        render: (v) => v.name,
      },
    ],
    dataSource: [
      {
        id: 1,
        name: 'jw',
      },
      {
        id: 2,
        name: 'jiangweixian',
      },
    ],
  })
}
