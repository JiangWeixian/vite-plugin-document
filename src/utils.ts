import path from 'node:path'
import { existsSync } from 'node:fs'

const DOCUMENT_FILES = ['Document.mjs', 'Document.cjs', 'Document.js']

export const resolveDistDocumentPath = (outDir: string, filenames = DOCUMENT_FILES) => {
  return filenames
    .map((filename) => path.resolve(outDir, filename))
    .find((filepath) => {
      return existsSync(filepath)
    })
}
