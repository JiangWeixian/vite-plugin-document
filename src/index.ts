import { build as viteBuild } from 'vite'
import type { Plugin } from 'vite'
import path from 'path'
import fs from 'fs'
import { renderToStaticMarkup } from 'react-dom/server'
import { resolveDistDocumentPath } from './utils'
import debug from 'debug'

let content = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/src/favicon.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/main.tsx"></script>
  </body>
</html>
`
const docType = (content: string) => `<!DOCTYPE html>${content}`

const VIRTUAL_HTML_ID = 'virtual-document.tsx'
const NAME = 'vite-plugin-document'

const log = {
  watcher: debug(`${NAME}:watcher`),
}

type Options = {
  documentFilePath?: string
  outDir?: string
}

export const VitePluginDocument = ({
  documentFilePath = '',
  outDir = '.vite-document',
}: Options = {}): Plugin => {
  const options: Required<Options> = {
    documentFilePath,
    outDir,
  }
  return {
    name: NAME,
    configResolved(config) {
      options.documentFilePath = documentFilePath || path.resolve(config.root, 'src/Document.tsx')
      options.outDir = path.resolve(config.root, options.outDir)
      if (!fs.existsSync(options.documentFilePath)) {
        console.error(`Document.tsx at ${options.documentFilePath} not exit!`)
      }
    },
    configureServer(server) {
      server.watcher.add(options.documentFilePath)
      server.watcher.on('all', async (_, filename) => {
        if (filename === options.documentFilePath) {
          log.watcher('%s changed full-reload', filename)
          const module = server.moduleGraph.getModuleById(VIRTUAL_HTML_ID)
          module && server.moduleGraph.invalidateModule(module!)
          if (server.ws) {
            server.ws.send({
              type: 'full-reload',
              path: '*',
            })
          }
        }
      })
      return () => {
        server.middlewares.use(async (req, res) => {
          const doc = (await server.ssrLoadModule(VIRTUAL_HTML_ID)).default
          content = docType(renderToStaticMarkup(doc()))
          content = await server.transformIndexHtml?.(req.originalUrl!, content, req.originalUrl)
          res.end(content)
        })
      }
    },
    resolveId(id) {
      if (id === VIRTUAL_HTML_ID) {
        return options.documentFilePath
      }
      if (id.endsWith('.html')) {
        return id
      }
      return id
    },
    async load(id) {
      // in build mode, vite will try to load `index.html`
      if (id.endsWith('.html')) {
        await viteBuild({
          logLevel: 'error',
          build: {
            outDir: options.outDir,
            ssr: options.documentFilePath,
          },
        })
        const distDocumentFilePath = resolveDistDocumentPath(options.outDir)!
        const doc = await import(distDocumentFilePath).then((res) => res.default)
        content = docType(renderToStaticMarkup(doc()))
        return {
          code: content,
          // https://rollupjs.org/guide/en/#source-code-transformations
          map: null,
        }
      }
      return null
    },
  }
}
