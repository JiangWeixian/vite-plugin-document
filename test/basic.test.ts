import { describe, it, expect } from 'vitest'
import { build, createServer } from 'vite'
import { VitePluginDocument } from '../src/index'
import path from 'node:path'
import { existsSync } from 'fs'

const f = (filepath = '') => path.resolve(__dirname, './fixtures', filepath)

describe('basic', () => {
  it('shout emit dist document filepath', async () => {
    await build({
      root: f('basic'),
      logLevel: 'silent',
      plugins: [
        VitePluginDocument({
          documentFilePath: f('basic/Document.tsx'),
        }),
      ],
    })
    const exit = existsSync(f('basic/.vite-document/Document.mjs'))
    expect(exit).toBe(true)
  })

  // it('document snapshot', async () => {
  //   const server = await createServer({
  //     root: f('basic'),
  //     logLevel: 'silent',
  //     plugins: [
  //       VitePluginDocument({
  //         documentFilePath: f('basic/Document.tsx'),
  //       }),
  //     ],
  //   })
  //   const result = await server.transformRequest('index.html')
  //   expect(result).toBeDefined()
  //   expect(result!.code).toMatchSnapshot('Document.tsx')
  // })

  it('options: { outDir: .document }', async () => {
    await build({
      root: f('basic'),
      logLevel: 'silent',
      plugins: [
        VitePluginDocument({
          documentFilePath: f('basic/Document.tsx'),
          outDir: '.document',
        }),
      ],
    })
    const exit = existsSync(f('basic/.document/Document.mjs'))
    expect(exit).toBe(true)
  })
})
