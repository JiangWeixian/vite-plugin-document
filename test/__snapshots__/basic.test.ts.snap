// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic > document snapshot > Document.tsx 1`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><meta charSet=\\"UTF-8\\"/><link rel=\\"icon\\" type=\\"image/svg+xml\\" href=\\"/src/favicon.svg\\"/><meta name=\\"env.mode\\" content=\\"prod\\"/><meta name=\\"env.prod\\" content=\\"prod\\"/><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\"/></head><body><div id=\\"root\\"></div></body></html>"`;
